Add support for rpm packaging

diehlpk opened the pull request:
Add support to generate rpm packages for HPX.

## Proposed Changes

  - Add HPX_WITH_RPM to generate a rpm packages using the cpack package
  - make package target generates the rpm packages containing the current HPX version
  -

## Any background context you want to provide?

make package generates hpx-1.2.0-1.x86_64.rpm in the build folder. The rpm packages contains boost, hwloc, and gperftools as its dependencies. 

sudo dnf install hpx-1.2.0-1.x86_64.rpm  install the headers, documentation, and all examples to /usr/local and installs the missing dependencies if they are not installed.

The package was generated on my desktop and installed within a VM using the same version of Fedora 28 and I was able to run the examples on the VM.

Once we have the debian package as well, the next step would be to integrate the packaging into our build process on circle-ci and store them as artifacts.

sithhell commented on .circleci/config.yml:
```diff
@@ -95,9 +95,14 @@ jobs:
                 -DCMAKE_CXX_FLAGS="-fcolor-diagnostics" \
                 -DHPX_WITH_TESTS_HEADERS=On \
                 -DHPX_WITH_DEPRECATION_WARNINGS=Off \
+<<<<<<< HEAD
```
This looks like a bad merge/rebase.

sithhell commented on CMakeLists.txt:
```diff
@@ -1989,6 +1989,12 @@ if(HPX_WITH_VIM_YCM)
 
+################################################################################
+# Add rpm packaging
+if(HPX_WITH_RPM)
```
Why is not enabled by default?

diehlpk replied:
I was thinking that this is a special use case and if one will build the package, one should explicitly enable it. I am open to enable it as default too.

sithhell replied:
Since it is just a target which has to be initiated explicitly, I don't see a problem having it optionally. If it however requires `rpm` or anything else to be installed, it needs to stay optional. Ideally however, it would be even nicer, if it could be implemented completely outside of our CMakeLists.txt infrastructure. I have no idea what that all entails. There used to be OpenSUSE packages here: https://build.opensuse.org/package/show/Education/hpx

No idea what happened there.

diehlpk replied:
Yes, one would need to install packages to build the rpm or debian package. Therefore, it should be optional. Let me get into contact with the maintainer for the opensuse packages.

junghans replied:
I made a quick update of the OpenSUSE package: https://build.opensuse.org/package/show/home:cjunghans:branches:Education/hpx

junghans replied:
Here is v1.1.0 https://build.opensuse.org/package/show/science:HPC/hpx

diehlpk commented:
Yes, it will work on any rpm-based distribution, but was only tested on Fedora since I am using fedora, but the rpm package does not depend on Fedora. 

Yes, you can generate rpms on debian systems by installing the rpm package.

junghans commented:
I did some rpm builds of 1.2.0-rc1 on Opensuse, which have different issue, see #3509, #3510, #3511.

hkaiser commented:
> I did some rpm builds of 1.2.0-rc1 on Opensuse, which have different issue, see #3509, #3510, #3511 and #3512. These platforms should be supported for an rpm package.

Thanks @junghans! I'm not sure however if we will be able to support all of the mentioned platforms. Some of the problems you reported (thanks!) are easily fixable, though.

junghans commented:
@hkaiser also see: https://github.com/diehlpk/hpx_packaging/issues/3

msimberg commented:
@diehlpk do you think we can close this? The official spec that you've worked on for Fedora is not generated (directly) by CPack. The only remaining use case would be sysadmins wanting to customize the spec for their system but considering the time you've put on the specs they're probably better off starting with your official spec anyway. Any other reason to include this in our repo?

Maybe you could update the Fedora instructions in our documentation?

junghans commented:
OpenSUSE is in review too: https://build.opensuse.org/request/show/650301

diehlpk commented:
> @diehlpk do you think we can close this? The official spec that you've worked on for Fedora is not generated (directly) by CPack. The only remaining use case would be sysadmins wanting to customize the spec for their system but considering the time you've put on the specs they're probably better off starting with your official spec anyway. Any other reason to include this in our repo?
> 
> Maybe you could update the Fedora instructions in our documentation?

The offical spec file build hpx with gcc, openmpi, and mpich. maybe someone does not want to have all this builds or want to enable papi or other packages in his rpm. 

I would prefer to keep it in the CMake file. it does not hurt if it is there or? 

I like to take care of it, when it would need updates in future.

diehlpk commented:
> Yeah, hpx should fit Fedora 28 and 29 soon, please test it with:
> 
> ```
> dnf install hpx --enablerepo=updates-testing
> ```

On Fedora 29 it was not available, but I will test it tomorrow again.

junghans commented:
> > Yeah, hpx should fit Fedora 28 and 29 soon, please test it with:
> > ```
> > dnf install hpx --enablerepo=updates-testing
> > ```
> 
> On Fedora 29 it was not available, but I will test it tomorrow again.

Yeah, it is still pending: https://bodhi.fedoraproject.org/updates/FEDORA-2018-a9b53adbdc

msimberg commented:
> I would prefer to keep it in the CMake file. it does not hurt if it is there or?

No, I don't think so. If you think it's useful it can definitely go in!

msimberg commented on cmake/packaging/rpm/Changelog.txt:
```diff
@@ -0,0 +1,2 @@
+* Wed Oct 24 2018 STE||AR Group <contact@stellar-group.org> 1.2.0-1
+- HPX Release 1.2
```
Does this need to be updated for every release? If so, could you add it to the release procedure?

hkaiser replied:
Could we add a link to some (permanent) webpage that refers to the latest release notes?

diehlpk replied:
@hkaiser Is this sufficient?
```

* Wed Oct 24 2018 STE||AR Group <contact@stellar-group.org> 1.2.0-1
- HPX Release 1.2 (https://stellar-group.github.io/hpx/docs/sphinx/tags/1.2.0/html/releases/whats_new_1_2_0.html)
```

diehlpk replied:
@msimberg Would be great to add it to the release procedure and update the version number and the link to the release notes.

junghans commented:
> > > Yeah, hpx should fit Fedora 28 and 29 soon, please test it with:
> > > ```
> > > dnf install hpx --enablerepo=updates-testing
> > > ```
> > 
> > 
> > On Fedora 29 it was not available, but I will test it tomorrow again.
> 
> Yeah, it is still pending: https://bodhi.fedoraproject.org/updates/FEDORA-2018-a9b53adbdc

It is in testing now.

diehlpk commented:
> > > > Yeah, hpx should fit Fedora 28 and 29 soon, please test it with:
> > > > ```
> > > > dnf install hpx --enablerepo=updates-testing
> > > > ```
> > > 
> > > 
> > > On Fedora 29 it was not available, but I will test it tomorrow again.
> > 
> > 
> > Yeah, it is still pending: https://bodhi.fedoraproject.org/updates/FEDORA-2018-a9b53adbdc
> 
> It is in testing now.

I installed all three different versions of the example package and tested the hello_world example with mpirun for openmpi and mpich and it was working. I will later compile a small hpx application with the devel package.

hkaiser commented on cmake/packaging/rpm/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+    /usr/local/lib
+  )
+set(CPACK_RPM_CHANGELOG_FILE "${CMAKE_HOME_DIRECTORY}/cmake/packaging/rpm/Changelog.txt")
+include(CPack)
```
This file misses and EOF.

junghans commented:
OpenSuse is done, now we just have to wait for Leap:15.1 integration: https://build.opensuse.org/request/show/652750

junghans commented:
And Fedora has two more days till stable:
https://bodhi.fedoraproject.org/updates/FEDORA-2018-9da04dcdf5
https://bodhi.fedoraproject.org/updates/FEDORA-2018-282193eb69

hkaiser commented:
@junghans, @diehlpk Can this be closed now?

junghans commented:
I think this isn't merged yet!

msimberg commented:
https://github.com/diehlpk/hpx/pull/1 is also still open.

msimberg commented on CMakeLists.txt:
```diff
@@ -1960,6 +1960,12 @@ if(HPX_WITH_VIM_YCM)
 
+################################################################################
+# Add rpm packaging
+if(HPX_WITH_RPM)
```
`HPX_WITH_RPM` should probably still be added with `hpx_option` in the main `CMakeLists.txt`.

diehlpk replied:
Add the hpx_option in the latest commit

diehlpk commented:
> I think this isn't merged yet!

Yes, I was busy with some administrative related things, which had to be finished soon. I will work on this today.

diehlpk commented:
> I think this isn't merged yet!

Yes, I was busy with some administrative related things, which had to be finished soon. I will work on this today.

diehlpk commented:
> I think this isn't merged yet!

Yes, I was busy with some administrative related things, which had to be finished soon. I will work on this today.

diehlpk commented:
@msimberg The latest commit should address your review

diehlpk commented:
@msimberg I cleaned up and added the pgp key again

msimberg commented:
Thanks @diehlpk! Looks good except that the addition to `release_procedure.rst` is now missing... If you add that back in I'll merge this.

diehlpk commented:
> Thanks @diehlpk! Looks good except that the addition to `release_procedure.rst` is now missing... If you add that back in I'll merge this.

@msimberg I added it again, it somehow got lost when I squashed my commits.
